Source/DafnyLanguageServer/Workspace/NotificationPublisher.cs

```diff
@@ -74,37 +79,80 @@ public class NotificationPublisher : INotificationPublisher {
     }
 
     private void PublishDocumentDiagnostics(IdeState previousState, IdeState state) {
-      var diagnosticParameters = GetPublishDiagnosticsParams(state);
-      var previousParams = GetPublishDiagnosticsParams(previousState);
-      if (previousParams.Version > diagnosticParameters.Version ||
-          previousParams.Diagnostics.SequenceEqual(diagnosticParameters.Diagnostics)) {
-        return;
+      var previousStateDiagnostics = previousState.GetDiagnostics();
+      var currentDiagnostics = state.GetDiagnostics();
+      if (state.Compilation.Project.IsImplicitProject) {
+        PublishImplicitProjectDiagnostics(state, currentDiagnostics, previousStateDiagnostics);
+      } else {
+        var uris = previousStateDiagnostics.Keys.Concat(currentDiagnostics.Keys).Distinct();
+
+        foreach (var uri in uris) {
+          IEnumerable<Diagnostic> current = currentDiagnostics.GetOrDefault(uri, Enumerable.Empty<Diagnostic>);
+          IEnumerable<Diagnostic> previous = previousStateDiagnostics.GetOrDefault(uri, Enumerable.Empty<Diagnostic>);
+          if (previous.SequenceEqual(current)) {
+            continue;
+          }
+
+          languageServer.TextDocument.PublishDiagnostics(new PublishDiagnosticsParams {
+            Uri = uri,
+            Version = filesystem.GetVersion(uri),
```

Comment:
There's two different versions. There's the file version, one for each file, what is what the client cares about because it will use that to do client side migration of positions. The client decides when to update the file versions. 

There's also the compilation version, which is irrelevant to the client, but something the server can use to disregard it's own outdated notifications. This used to be the same version as the file version since there was only a single file triggering compilations, but now it's a separate integer. Maaaybe we could get rid of it but this way the logic has changed the least.

Reply:
Thanks for the explanation, that makes sense.